wp-content/plugins/tapestry/controller/class.tapestry-controller.php

```diff
@@ -3,4 +3,163 @@
+                $node->id = add_post_meta($postId, 'tapestry_node', $metadata);
+            } else {
+                $metadata = get_metadata_by_mid('post', $node->id)->meta_value;
+                $nodePostId = $metadata->post_id;
+            }
+            update_post_meta($nodePostId, 'tapestry_node_data', $node);
+        }
+    }
+
+    private function makeMetadata($node, $nodePostId) {
+        return (object) array(
+            'post_id' => $nodePostId,
+            'title' => $node->title,
+            'coordinates' => (object) array(
+                'x' => $node->fx,
+                'y' => $node->fy
+            )
+        );
+    }
+    
+    private function getTapestryById($postId) {
+        $tapestry = get_post_meta($postId, 'tapestry', true);
+        $metadatas = array_map(function($nodeId) {
+            return get_metadata_by_mid('post', $nodeId);
+        }, $tapestry->nodes);
+        $nodeDatas = array_map(function($metadata) {
+            $nodeData = get_post_meta($metadata->meta_value->post_id, 'tapestry_node_data', true);
+            return $this->updateNodeData($nodeData, $metadata);
+        }, $metadatas);
+
+        $tapestry->nodes = $nodeDatas;
+
+        // TODO: delete the below when being able to create tapestry from scratch
+        $tapestry->links = $this->getNewLinks($tapestry->links, $nodeDatas);
+        return $tapestry;
+    }
+
+    private function updateNodeData($nodeData, $metadata) {
+        // Update node data here to match its own version
+        // This enables the same node to have multiple versions
+        $nodeData->id = (int) $metadata->meta_id;
+        $nodeData->title = $metadata->meta_value->title;
+        $nodeData->fx = $metadata->meta_value->coordinates->x;
+        $nodeData->fy = $metadata->meta_value->coordinates->y;
+        return $nodeData;
+    }
+
+    private function getNodeIds($nodes) {
+        return array_map(function($node) {
+            return $node->id;
+        }, $nodes);
+    }
+
+    private function insertPost($post, $type) {
+        switch($type) {
+            case self::POST_TYPES['TAPESTRY_NODE']:
+                $postType = $post->type;
+                $postTitle = $post->title;
+                $postStatus = $post->status;
+                break;
+            case self::POST_TYPES['TAPESTRY']:
+            default:
+                $postType = $post->settings->type;
+                $postTitle = $post->settings->title;
+                $postStatus = $post->settings->status;
+                break;
+        }
+        return wp_insert_post(array(
+            'post_type' => $postType,
+            'post_status' => $postStatus,
+            'post_title' => $postTitle
+        ), true);
+    }
+
+    // TODO: Remove this when we can build a tapestry from scratch
+    // HACK - create a new links array that works with new IDs
+    private function getNewLinks($oldLinks, $nodes) {
```

Comment:
Similarly, this should also be moved to the other PR.

Reply:
I removed it.